feat: names for local plugins

zcuric opened the pull request:
<!-- Please don't delete this template -->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [x] Docs
- [ ] Underlying tools
- [ ] Other, please describe:

<!--
Note:
When submitting documentation PRs, please target the `master` branch (https://cli.vuejs.org) or `next` branch (https://next.cli.vuejs.org)
When submitting coding PRs, please target the `dev` branch.
-->

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

**Other information:**
This PR tries to simplify getting plugin options when developing local plugins, utilising both the`vuePlugins.service` field in the `package.json` and `pluginOptions` in the `vue.config.js`.
Using `Object` instead of an `Array` for plugin registration, for example: 
```json
{
  "vuePlugins": {
    "service": {
      "myCommands": "my-commands.js"
    }
  }
}
```
key `myCommands` becomes the plugin `name`. If that same `name` is used as key in `pluginOptions` field, for example: 
```js
{
...
  "pluginOptions": {
    "myCommands": {
      "foo": "bar",
      "baz": "quz"
    }
  }
...
}
```
options for that plugin will become available as the third argument in the `PluginAPI`, for example: 
```js
// my-commands.js
module.exports = (api, options, { foo, baz }) => {
  console.log(foo) // bar
  console.log(baz) // quz
}
```

Also, this PRs tests now cover registration of local plugins in both cases.

pksunkara commented on docs/dev-guide/plugin-dev.md:
````diff
@@ -432,6 +432,49 @@ module.exports.defaultModes = {
+  }
+}
+```
+When using the object configuration, key `myCommands` becomes your local plugin `name` and the plugin options will be directly accessible as the third argument in `PluginAPI` if the plugin `name` corresponds to the key defined in `pluginOptions`.
````
There's no context for `myCommands` here.

zcuric replied:
You are correct.

zcuric commented:
> I don't understand why local plugins need a name when you can simply use `local:${file}`. 
To avoid: `local:plugins/my-super-cool-plugin`.

> You also added an unnecessary `options` to PluginAPI. 
Have your read the reasoning I've added this **optional** `options`? It's there to provide easier access to plugin options.

pksunkara commented:
> To avoid: `local:plugins/my-super-cool-plugin`.

Yeah, could you please expand on that a bit more? What were the issues you are having and how does this fix it? Etc..

zcuric commented:
> > To avoid: `local:plugins/my-super-cool-plugin`.
> 
> Yeah, could you please expand on that a bit more? What were the issues you are having and how does this fix it? Etc..

It provides easier access to plugins options if plugin is named. I found it cumbersome to extract the plugin options from the `service` when developing plugins. `poi` has this solved nicely by providing directly plugin options as an argument, so this was inspired by it. 

I wasn't having any issues, I just wanted to make things a bit more convenient for plugin developers. 
This doesn't have to be merged, it's just my proposal, although PR has some neat tests that cover existing functionality.

pksunkara commented:
> It provides easier access to plugins options if plugin is named. I found it cumbersome to extract the plugin options from the service when developing plugins. poi has this solved nicely by providing directly plugin options as an argument, so this was inspired by it.

Yeah, we can provide the pluginOptions as an argument. But, it should be done all kinds of plugins, not only local ones.

It still doesn't change the fact that we don't have to add support for `name` for local plugins. The code can be changed:

```diff
-      const options = pluginOptions[name] || {}
+      const options = pluginOptions[id] || {}
```

zcuric commented:
> it still doesn't change the fact that we don't have to add support for name for local plugins. The code can be changed

I see your point. I'll update.

zcuric commented:
@pksunkara Closing this one. I'll make a new PR.
